modules/vault/src/main/java/org/apache/ignite/internal/vault/persistence/PersistentVaultService.java

```diff
@@ -0,0 +1,207 @@
+
+    /** */
+    private final Options options = new Options();
+
+    /** */
+    private final RocksDB db;
+
+    /**
+     * Creates and starts the RocksDB instance using the recommended options on the given {@code path}.
+     *
+     * @param path base path for RocksDB
+     */
+    public PersistentVaultService(Path path) {
+        // using the recommended options from https://github.com/facebook/rocksdb/wiki/Setup-Options-and-Basic-Tuning
+        options
+            .setCreateIfMissing(true)
+            .setCompressionType(CompressionType.LZ4_COMPRESSION)
+            .setBottommostCompressionType(CompressionType.ZSTD_COMPRESSION)
+            .setLevelCompactionDynamicLevelBytes(true)
+            .setBytesPerSync(1048576)
+            .setCompactionPriority(CompactionPriority.MinOverlappingRatio)
+            .setTableFormatConfig(
+                new BlockBasedTableConfig()
+                    .setBlockSize(16 * 1024)
+                    .setCacheIndexAndFilterBlocks(true)
+                    .setPinL0FilterAndIndexBlocksInCache(true)
+                    .setFormatVersion(5)
+                    .setFilterPolicy(new BloomFilter(10, false))
+                    .setOptimizeFiltersForMemory(true)
+            );
+
+        try {
+            db = RocksDB.open(options, path.toString());
+        }
+        catch (RocksDBException e) {
+            throw new IgniteInternalException(e);
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public void close() throws RocksDBException {
+        try (options; db) {
+            db.syncWal();
+
+            IgniteUtils.shutdownAndAwaitTermination(threadPool, 10, TimeUnit.SECONDS);
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public @NotNull CompletableFuture<VaultEntry> get(@NotNull ByteArray key) {
+        return supplyAsync(() -> db.get(key.bytes()))
+            .thenApply(v -> new VaultEntry(key, v));
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public @NotNull CompletableFuture<Void> put(@NotNull ByteArray key, byte @Nullable [] val) {
+        return val == null ? remove(key) : runAsync(() -> db.put(key.bytes(), val));
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public @NotNull CompletableFuture<Void> remove(@NotNull ByteArray key) {
+        return runAsync(() -> db.delete(key.bytes()));
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public @NotNull Cursor<VaultEntry> range(@NotNull ByteArray fromKey, @NotNull ByteArray toKey) {
+        try (var readOpts = new ReadOptions()) {
+            var lowerBound = new Slice(fromKey.bytes());
+            var upperBound = new Slice(toKey.bytes());
+
+            readOpts
+                .setIterateLowerBound(lowerBound)
+                .setIterateUpperBound(upperBound);
+
+            RocksIterator it = db.newIterator(readOpts);
+            it.seekToFirst();
+
+            return new RocksIteratorAdapter(it, lowerBound, upperBound);
```

Comment:
Should you pass readOpts as well? %)

Reply:
this is a very tricky question: if I don't pass the lower/upper bounds, I always get a Segmentation Fault (looks like they get closed too early). But closing the `readOpts` seems to always work, so I think we can leave it here

Reply:
Sure, not a problem at all